### ISSUE TITLE
Supernova seems only compatible with VS 2017 on Windows

### ISSUE THREAD

**Issue by balintlaczko** [supercollider/supercollider#5899]
## Environment

* SuperCollider version: both 3.12.2 and local build of 3.13.0-dev
* Operating system: Windows 10
* Other details (Qt version, audio driver, etc.): Qt 5.15.2, ASIO

## Steps to reproduce

```supercollider
Server.supernova;
s.boot;
```

## Expected vs. actual behavior

I get the infamous exit code -1073741676, which is [apparently a divide-by-zero error](https://stackoverflow.com/questions/47422982/c-status-1073741676).

I have seen some other people facing this issue too, like [here](https://www.reddit.com/r/supercollider/comments/wnylee/im_trying_to_boot_supernova_now_with_312/) and [here](https://scsynth.org/t/win11-supernova/6377).

I compiled SC from source (with Supernova), and it seems that strangely enough if I am building in Debug mode, Supernova manages to boot with some quirks:
- a) I get the [thread affinity warnings](https://github.com/supercollider/supercollider/issues/5540) that indeed go away if I build from [this PR](https://github.com/supercollider/supercollider/pull/5618).
- b) when I try to use the default Synth, I get an error: `OffsetOut_Ctor: alloc failed, increase server's RT memory (e.g. via ServerOptions)` which does not seem affected by the size of `Server.default.options.memSize`. However if I define a SynthDef, that seems to work okay.

SC 3.13.0-dev in Debug mode (this works):
```supercollider
Server.supernova;
s.boot;

(
SynthDef("sine", { arg freq=800, chan=0;
    var osc;
    osc = SinOsc.ar(freq, 0, mul:1/600);
    Out.ar(chan, osc);
}).writeDefFile;
)

(
var group = ParGroup.new();
600.do{ Synth(\sine, target: group) };
)
```

BUT if I try this (still in SC 3.13.0-dev in Debug):
```supercollider
Server.supernova;
s.boot;

(
var group = ParGroup.new();
n = 10
n.do{ Synth(\default, target: group) };
)
```
I get a `OffsetOut_Ctor: alloc failed, increase server's RT memory (e.g. via ServerOptions)` n times.

If I build in Release mode, Supernova refuses to boot with the familiar exit code -1073741676, again.
SC 3.12.2 (public release) OR 3.13.0-dev built in Release mode:
```supercollider
Server.supernova;
s.boot;
```

I get:
```
Booting server 'localhost' on address 127.0.0.1:57110.
Supernova booting
Server 'localhost' exited with exit code -1073741676.
```

So there is something with building in Release mode that produces a divide-by-zero error when Supernova tries to boot. It seems to work (minus the default synth alloc error) if I build in Debug mode.

**Comment by dyfer** [comment]
Hello @balintlaczko 
What are you server options? Have you tried changing e.g. ASIO buffer size? 
Also, supernova tries to boot at 44100 IIRC (it does _not_ take the samplerate from hardware), have you tried setting that parameter explicitly in ServerOptions to match your hardware?

> I get the infamous exit code -1073741676, which is [apparently a divide-by-zero error](https://stackoverflow.com/questions/47422982/c-status-1073741676).

Often this kind of error is the PortAudio (audio backend) error that bubbles up.

**Comment by balintlaczko** [comment]
Here is my startup file that I used in all cases when testing:

```supercollider
o = Server.default.options;
o.sampleRate_(48000);
o.blockSize_(128);
o.device = "ASIO : Focusrite USB ASIO";
o.memSize = 1024 * 1024;
o.numAudioBusChannels = 8192;
o.numBuffers = 8192;
o.maxNodes = 8192;
```

> Also, supernova tries to boot at 44100 samplerate IIRC (it does not take the samplerate from hardware on Windows/ASIO)
I was totally unaware of this, let me test and I'll report back here! Thanks for the tip!

**Comment by balintlaczko** [comment]
Nope, testing with 3.12.2-release, it is the same on both 44100/48000 and 128/512. I matched the hardware settings in the server options in each case.

**Comment by balintlaczko** [comment]
The strange thing to me is that the same thing boots and works in a debug build and fails to boot in the release build.

**Comment by dyfer** [comment]
> I was totally unaware of this, let me test and I'll report back here! Thanks for the tip!

You're welcome! Just to clarify, if you set sample rate with ServerOptions, it will try to set the ASIO device to that sample rate. If the device supports that sample rate, this should work fine. When you don't specify the samplerate, it will try to boot to the first samplerate reported as "supported" by the hardware from the list: 44100, 48000 etc. So this fails mostly for hardware where sample rate is set externally and can't be changed in software (e.g. with a switch, wordclock etc). This is PortAudio's behavior.

> The strange thing to me is that the same thing boots and works in a debug build and fails to boot in the release build.

Yeah, that's strange. I don't know what's going on.

**Comment by balintlaczko** [comment]
Good news! Managed to compile [this PR](https://github.com/supercollider/supercollider/pull/5618) in Release mode with Visual Studio 15 2017 build tools. Previously I compiled with Visual Studio 2022, with that it still does not work (in release).
I still get the `OffsetOut_Ctor: alloc failed, increase server's RT memory (e.g. via ServerOptions)` error whenever I try to use the default Synth, otherwise, things look good!
![image](https://user-images.githubusercontent.com/45127463/200071477-bd2bd3bf-8fe1-4ce9-80a0-6b109bb1357f.png)

But have to test more to see if everything works as should.

**Comment by dyfer** [comment]
> Previously I compiled with Visual Studio 2022, with that it still does not work (in release).

In that case this seems to be an issue of compatibility with Visual Studio 2022. In the CI we only build with Visual Studio 16 2019.

If you confirm that local build with VS 2017 works as expected, I'd suggest either renaming the title of this issue, or closing this one and filing new one mentioning an issue when building with VS 2022.

**Comment by balintlaczko** [comment]
Actually, I noticed now that I get the same exit code with Supernova if I am using the VS 2019 build tools (scsynth still works). Going back to targeting VS 2017 supernova boots and works again. So supernova only seems to be compatible with VS 2017.

**Comment by dyfer** [comment]
SC is built using VS 2019 in GitHub Actions.
Could you check if you experience this problem on a [recent develop build from github actions](http://supercollider.s3.amazonaws.com/builds/supercollider/supercollider/win64/develop-latest.html)?

BTW which version of libsndfile are you using when building SC yourself?

**Comment by balintlaczko** [comment]
Yes, the build you linked gives me the same exit code when trying to boot Supernova.
In the local builds I am using libsndfile 1.0.28 (seems to be the latest(?) at the time of writing).
Both when I compile from [this PR](https://github.com/supercollider/supercollider/pull/5618) and from the develop branch, Supernova only is able to boot when I use VS 2017 build tools. Scsynth works fine in all toolsets (VS 2017-2019-2022).
When using the develop branch (compiled with VS 2017) I get the expected [thread affinity warnings](https://github.com/supercollider/supercollider/issues/5540) that won't show up if I use the PR branch instead. But Supernova indeed boots in both, only when in VS 2017.

However, both on the PR and the develop branches (when compiled with VS 2017) AND booted with Supernova, when I try to use the default synth I get this error:
```supercollider
Synth(\default, [\freq, 440, \amp, 0.5])
```
```
OffsetOut_Ctor: alloc failed, increase server's RT memory (e.g. via ServerOptions)
```
When using Scsynth the default synth works. Should I open a separate issue for this?

**Comment by dyfer** [comment]
Thank you for confirming and checking other builds.

I was able to reproduce both issues using the latest develop build from our build system. **It seems that supernova can't boot only if a device name is set** - it does boot for me if I don't set it. Of course this is not a "fix", but something to keep in mind when trying to debug this.

I can also reproduce the `OffsetOut_Ctor` error.

No need to open a separate issue, at least not until we find the reason for at least one of them. Unfortunately, I don't know what might be wrong. @Spacechild1 do you have any ideas how to debug this?

**Comment by dyfer** [comment]
Thanks so much @Spacechild1 ! I'll test this build.
However, do you have any idea why this only happens with MSVC 2019 and up, and not with gcc/clang and/or other platforms?

**Comment by balintlaczko** [comment]
> It seems that supernova can't boot only if a device name is set

I cannot confirm this with the develop branch compiled via VS 2019. Supernova either does not boot with some devices, or hangs for a long time (NB. windows defender RT protection is off) and then quits. This also does not seem to be consistent (sometimes it hangs with the same device, then quits, sometimes it quits immediately - could it be that something gets "dirty" when trying with device A and then remains "dirty" when trying with device B?).

Thanks so much for #5902 , @Spacechild1 ! It works for me!

Test build via VS 2019:
- supernova doesn't boot (expected due to VS 2019)

Test build via VS 2017:
- supernova boots, with thread affinity warnings (expected, and can be fixed by #5618)
- upon trying to use the default Synth the `OffsetOut_Ctor` error is gone, and all seems to be in order :)

So what remains a mystery is why MSVC 2019 has a problem where MSVC 2017 doesn't. Maybe something got deprecated?

**Comment by Spacechild1** [comment]
> Thanks so much for https://github.com/supercollider/supercollider/pull/5902 , @Spacechild1 ! It works for me!

Great! @dyfer #5902 is now ready for review/merging.

---

> So what remains a mystery is why MSVC 2019 has a problem where MSVC 2017 doesn't. 

I'm travelling at the moment, but once I get back to my Windows dev machine I will have a look.

**Comment by Spacechild1** [comment]
@balintlaczko In the meantime you could use some good ol' printf-debugging to see how far the `supernova.exe` program gets before it decides to terminate. (Or run it in a debugger with a few breakpoints.)

**Comment by dyfer** [comment]
> > It seems that supernova can't boot only if a device name is set
> 
> I cannot confirm this with the develop branch compiled via VS 2019. Supernova either does not boot with some devices, or hangs for a long time (NB. windows defender RT protection is off) and then quits. This also does not seem to be consistent (sometimes it hangs with the same device, then quits, sometimes it quits immediately - could it be that something gets "dirty" when trying with device A and then remains "dirty" when trying with device B?).

Thanks for checking this.
I was trying to test https://github.com/supercollider/supercollider/pull/5902 and indeed I encountered a similarly inconsistent behavior - even with an older build where supernova used to boot before (when no device name was set). So I now thing that my comment was inaccurate and device name setting is a red herring.

@Spacechild1 thanks for much for your help and fixes!

**Comment by balintlaczko** [comment]
Again, I really don't think I know what I'm doing, but I may have narrowed it down to either of three lines in [server.cpp](https://github.com/supercollider/supercollider/blob/develop/server/supernova/server/server.cpp):
- [scheduler<thread_init_functor>(args.threads, !args.non_rt),](https://github.com/supercollider/supercollider/blob/02c37d28badc390f7ee320883b955052294d3cd6/server/supernova/server/server.cpp#L52) OR
- [buffer_manager(args.buffers),](https://github.com/supercollider/supercollider/blob/02c37d28badc390f7ee320883b955052294d3cd6/server/supernova/server/server.cpp#L53) OR
- [sc_osc_handler(args)](https://github.com/supercollider/supercollider/blob/02c37d28badc390f7ee320883b955052294d3cd6/server/supernova/server/server.cpp#L54)
- 
I could be wrong, I tried to printf everything, but it all goes silent after `server_shared_memory_creator` finishes (the line before the scheduler). And it never seems to reach line 55.  I hope I'm not just stating the obvious and this might be useful for ppl with more iq points. :)

**Comment by Spacechild1** [comment]
@balintlaczko That's good progress! Now you could dive into the `server_shared_memory_creator` and continue recursively until you've found the exact place where the program aborts. Then I might be able to spot the problem.

**Comment by bobvodka** [comment]
As I've also hit this issue I've taken a look at it, and tracked the issue down to [ void calibrate_backoff(int timeout_in_seconds)](https://github.com/supercollider/supercollider/blob/d52b0522b210def4fe47829ca332f0bc411b3509/server/supernova/dsp_thread_queue/dsp_thread_queue.hpp#L421) specifically the line;

[watchdog_iterations = (seconds(timeout_in_seconds) / median) * backoff_iterations;] (https://github.com/supercollider/supercollider/blob/d52b0522b210def4fe47829ca332f0bc411b3509/server/supernova/dsp_thread_queue/dsp_thread_queue.hpp#L444)

_median_ is coming out as 0, causing a divide by zero error and the explosion.

_median_ is being pulled from the middle of an array in that function, which has 16 timing values generated in it - however 13 of them (on this debug run I'm looking at on my machine) are 0. 

Making _median_ = 100ns let Supernova boot correctly (although clearly not a robust solution), and was able to correctly get sound out of it using a ParGroup to contain 100 instances of the default instrument. (Rebooted and tested with 100 instances outside of a ParGroup just to check the load was lower, which it was, so seems good to me.)

For the record, my system;
- Windows 10
- Ryzen 9 5950X 16 core @ 3.4Ghz/4.0Ghz
- Built using Visual Studio 22, using CMake configured for VS 2019 x64

**Comment by Spacechild1** [comment]
@bobvodka Good catch! Maybe VS 2019 changed the underlying clock function for `std::high_resolution_clock`? In general, Windows system clock is not particularly precise. I think we should rather use https://learn.microsoft.com/en-us/windows/win32/api/profileapi/nf-profileapi-queryperformancecounter for calibrating the backoff.

Or course, we should still handle the case where median is 0, for example:
```
auto median = measured_values[measured_values.size() / 2];
if (median > 0) {
    watchdog_iterations = (seconds(timeout_in_seconds) / median) * backoff_iterations;
} else {
    std::cout << "WARNING: backoff calibration failed, watchdog disabled\n";
    watchdog_iterations = std::numeric_limits<int>::max();
}
```

**Comment by bobvodka** [comment]
It seems like the Windows/MSVC implementation of `std::high_resolution_clock` switched to using `queryperformancecounter` back in 2015 (well, `high_resolution_clock` is an alias for `steady_clock` which uses QPC) - so it should be giving the best resolution possible.

I think what has likely happened is the optimiser has improved since 2017, as looking at the disassembly for the function, the loop for calling `backoff::run` simply doesn't exist; tracing it you get a call to `chrono::steady_clock::now()` followed directly by another call to it and then the maths to see how long has passed. 

So, that would explain the problem - the values coming out are basically noise due to a change in optimisations, which would explain why the values I've seen have had mostly zero in, and also why the debug version works as that will keep that code in place.

**Comment by Spacechild1** [comment]
> It seems like the Windows/MSVC implementation of std::high_resolution_clock switched to using queryperformancecounter back in 2015 (well, high_resolution_clock is an alias for steady_clock which uses QPC) - so it should be giving the best resolution possible.

That's good to know!

>  the loop for calling backoff::run simply doesn't exist;

Ouch! I had a quick look and I think the problem is the implementation of `nova::detail::pause`:
https://github.com/timblechmann/nova-tt/blob/55da93741e76632579b63004d3378139e32d634f/nova-tt/pause.hpp#L31-L50

I haven't tested, but I think MSVC does not define the `__SSE2__` macro. (I could not find it in https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170&redirectedfrom=MSDN.) In that case, the `pause()` function would be empty and the compiler has every right to remove the function call.

(I also noticed that `pause` has no implementation for ARM!)

For comparison, here's my own implementation which seems to work with all compilers:

https://github.com/Spacechild1/aoo/blob/b2f1b583aafe7846a04aaf32e32467aac6593e0a/common/sync.hpp#L32-L77

I can make a PR to `nova-tt`. @timblechmann Any comments/suggestions?

**Comment by bobvodka** [comment]
Just did a quick test locally by changing the `#ifdef` to `#if 1` in the `pause.hpp` file to force the usage of `__mm_pause()` and, as expected, it did result in some sensible numbers appearing in all 16 entries for the timings, no more divide by zero 🥳

So I would second the idea of making that change to the code base 👍

**Comment by Spacechild1** [comment]
Thanks for testing! I will make a PR to nova-tt. Once that has been accepted, I'll make a PR to SuperCollider.

**Comment by Spacechild1** [comment]
@bobvodka Can you test https://github.com/supercollider/supercollider/pull/6402?

**Comment by bobvodka** [comment]
@Spacechild1 Yep, looks good to me; builds fine, timings in the array look sensible now for all 16 entries, and of course no divide by zero 👍

**Comment by Spacechild1** [comment]
Yay! Thanks for testing! @dyfer @muellmusik @JordanHendersonMusic Anyone interested in a review? :)
